cdap-data-fabric/src/test/java/co/cask/cdap/spi/metadata/MetadataStorageTest.java

```diff
@@ -0,0 +1,1101 @@
+    Assert.assertNull(response.getCursor());
+    // TODO (CDAP-14584) expect 3 here after this bug is fixed
+    // Assert.assertEquals(3, response.getTotalResults());
+
+    // clean up
+    mds.batch(ImmutableList.of(
+      new Drop(app), new Drop(dataset), new Drop(program)));
+  }
+
+  private void verifyMetadataSelection(MetadataStorage mds, MetadataEntity entity, Metadata metadata,
+                                       ImmutableSet<ScopedNameOfKind> selection) throws IOException {
+    Assert.assertEquals(filterBy(metadata, selection), mds.read(new Read(entity, selection)));
+  }
+
+  private void verifyMetadata(MetadataStorage mds, MetadataEntity entity, Metadata metadata) throws IOException {
+    // verify entire metadata
+    Assert.assertEquals(metadata, mds.read(new Read(entity)));
+
+    // filter by scope
+    verifyFilteredMetadata(mds, entity, metadata, SYSTEM, null);
+    verifyFilteredMetadata(mds, entity, metadata, USER, null);
+
+    // filter by kind
+    verifyFilteredMetadata(mds, entity, metadata, null, PROPERTY);
+    verifyFilteredMetadata(mds, entity, metadata, null, MetadataKind.TAG);
+
+    // filter by kind and scope
+    verifyFilteredMetadata(mds, entity, metadata, SYSTEM, PROPERTY);
+    verifyFilteredMetadata(mds, entity, metadata, SYSTEM, MetadataKind.TAG);
+    verifyFilteredMetadata(mds, entity, metadata, USER, PROPERTY);
+    verifyFilteredMetadata(mds, entity, metadata, USER, MetadataKind.TAG);
+  }
+
+  private void verifyFilteredMetadata(MetadataStorage mds, MetadataEntity entity, Metadata metadata,
+                                      MetadataScope scope, MetadataKind kind) throws IOException {
+    Assert.assertEquals(filterBy(metadata, scope, kind), mds.read(new Read(entity, scope, kind)));
+  }
+
+  private Metadata filterBy(Metadata metadata, MetadataScope scope, MetadataKind kind) {
+    //noinspection ConstantConditions
+    return new Metadata(
+      kind == PROPERTY ? Collections.emptySet()
+        : Sets.filter(metadata.getTags(), tag -> scope == null || scope.equals(tag.getScope())),
+      kind == MetadataKind.TAG ? Collections.EMPTY_MAP
+        : Maps.filterKeys(metadata.getProperties(), key -> scope == null || scope.equals(key.getScope())));
+  }
+
+  private Metadata filterBy(Metadata metadata, ImmutableSet<ScopedNameOfKind> selection) {
+    //noinspection ConstantConditions
+    return new Metadata(
+      Sets.filter(metadata.getTags(), tag ->
+        selection.contains(new ScopedNameOfKind(MetadataKind.TAG, tag.getScope(), tag.getName()))),
+      Maps.filterKeys(metadata.getProperties(), key ->
+        selection.contains(new ScopedNameOfKind(PROPERTY, key.getScope(), key.getName()))));
+  }
+
+
+  private static void assertEmpty(MetadataStorage mds, SearchRequest request) throws IOException {
+    Assert.assertTrue(mds.search(request).getResults().isEmpty());
+  }
+
+  protected static SearchResponse assertInOrder(MetadataStorage mds, SearchRequest request,
```

Comment:
Are these protected for devs to extend and use them while writing their own tests?

Reply:
yes, actually the next PR adds an implementation-specific test for the existing Dataset-based implementation, which uses these methods.